builder/steps/gen-dockerfile/contents/data/Dockerfile.txt

```diff
@@ -1,14 +1,52 @@
+<% if (config.buildCommand) { %>
+FROM <%- baseImage %> as build_step
+COPY . /app/
+<% if (config.canInstallDeps) { %>
+ENV NODE_ENV development
+  <% if (config.useYarn) { %>
+RUN yarn install || \
+  ((if [ -f yarn-error.log ]; then \
+      cat yarn-error.log; \
+    fi) && false)
+  <% } else { %>
+# You have to specify "--unsafe-perm" with npm install
+# when running as root.  Failing to do this can cause
+# install to appear to succeed even if a preinstall
+# script fails, and may have other adverse consequences
+# as well.
+# This command will also cat the npm-debug.log file after the
+# build, if it exists.
+RUN npm install --unsafe-perm || \
+  ((if [ -f npm-debug.log ]; then \
+      cat npm-debug.log; \
+    fi) && false)
+  <% } %>
+ENV NODE_ENV production
+<% } %>
+RUN <%- config.buildCommand %>
```

Comment:
For security/paranoia's sake, should you check that "config.buildCommand" doesn't have a newline or similar?  Basically I'm thinking the Bash/Docker equivalent of SQL injection attacks, like someone's app.yaml having `runtime_config: build: "true\\nCOPY /etc/passwd /pwned"`.